jetbrains-core/src/software/aws/toolkits/jetbrains/services/codewhisperer/telemetry/CodeWhispererCodeCoverageTracker.kt

```diff
@@ -0,0 +1,150 @@
+import software.aws.toolkits.jetbrains.services.codewhisperer.model.SessionContext
+import software.aws.toolkits.jetbrains.services.codewhisperer.popup.CodeWhispererPopupManager
+import software.aws.toolkits.jetbrains.services.codewhisperer.popup.CodeWhispererUserActionListener
+import software.aws.toolkits.jetbrains.services.codewhisperer.service.CodeWhispererService
+import software.aws.toolkits.telemetry.CodewhispererLanguage
+import software.aws.toolkits.telemetry.CodewhispererTelemetry
+import java.time.Duration
+import java.time.Instant
+import java.util.concurrent.atomic.AtomicBoolean
+import java.util.concurrent.atomic.AtomicInteger
+import kotlin.math.roundToInt
+
+abstract class CodeWhispererCodeCoverageTracker(
+    private val timeWindowInSec: Long,
+    private val language: CodewhispererLanguage,
+    acceptedTokensSize: AtomicInteger,
+    totalTokensSize: AtomicInteger
+) : Disposable {
+    val percentage: Int
+        get() = if (totalTokensSize.get() != 0) (acceptedTokensSize.get().toDouble() / totalTokensSize.get() * 100).roundToInt() else 0
+    var acceptedTokensSize = acceptedTokensSize
+        private set
+    var totalTokensSize = totalTokensSize
+        private set
+    private val alarm = AlarmFactory.getInstance().create(Alarm.ThreadToUse.POOLED_THREAD, this)
+    private val isShuttingDown = AtomicBoolean(false)
+    private var startTime: Instant = Instant.now()
+
+    init {
+        val conn = ApplicationManager.getApplication().messageBus.connect()
+        conn.subscribe(
+            CodeWhispererPopupManager.CODEWHISPERER_USER_ACTION_PERFORMED,
+            object : CodeWhispererUserActionListener {
+                override fun beforeAccept(states: InvocationContext, sessionContext: SessionContext) {
+                    // Should remove right paranthesis added by IDE
+                    val editor = states.requestContext.editor
+                    val typeahead = sessionContext.typeahead
+                    val originalOffset = editor.caretModel.primaryCaret.offset - typeahead.length
+                    val codewhispererMetadata = editor.getUserData(CodeWhispererService.KEY_CODEWHISPERER_METADATA)
+                    val endOffsetToReplace = codewhispererMetadata?.insertEnd ?: editor.caretModel.primaryCaret.offset
+                    val rightParanthesis = endOffsetToReplace - originalOffset
+                    totalTokensSize.addAndGet(-rightParanthesis)
+                }
+
+                override fun afterAccept(states: InvocationContext, sessionContext: SessionContext, remainingRecomm: String) {
+                    if (states.requestContext.fileContextInfo.programmingLanguage.toCodeWhispererLanguage() != language) return
+                    pushAcceptedTokens(remainingRecomm)
+                }
+            }
+        )
+        scheduleCodeWhispererTracker()
+    }
+
+    fun documentChanged(event: DocumentEvent) {
+        pushTotalTokens(event.newFragment)
```

Comment:
This only covers the case when the new text is *added* to the document. 
You may also want to consider the other `DocumentEvent` cases such as:
1. Removed the existing text (You can use `event.oldLength`)
2. Replaced the existing text (You need to use old length and new length.)

Or to generalize you can just add (`event.newLength - event.oldLength`) number of tokens to the `totalTokens` to cover all the cases. Although this won't work as the user may replace the text recommended by CodeWhisperer. 
I think this logic is much more involved and needs to be taken into consideration many other cases.